### PR TITLE
#35  Proposes an extension to access GATT descriptors

### PR DESCRIPTION
**Pull request by jlusiardi**
This proposes an extension to gatt-python to be able to read out
the value of GATT descriptors of characteristics as well. Those
descriptors can contain metadata and configuration information
for their characteristic.

This could fix my request in #35

### CODE REVIEW

**Review by larsblumberg — COMMENTED** [review]
Thanks for providing this MR, it looks great. I have one comment, can you please apply that?

File: gatt/gatt_linux.py [hunk 1]
```diff
@@ -521,6 +557,15 @@ def __init__(self, service, path, uuid):
...
 
+        # extract the descriptions from the paths obtained via dbus. They end with /descXXXX.
+        self.descriptors = []
+        descriptor_regex = re.compile(self._path + '/desc[0-9abcdef]{4}$')
+        managed_descriptors = [desc for desc in self._object_manager.GetManagedObjects().items() if
+                               descriptor_regex.match(desc[0])]
+
+        for desc in managed_descriptors:
+            self.descriptors.append(Descriptor(self, desc[0], desc[1]['org.bluez.GattDescriptor1']['UUID']))
+
```
[the remark this answers is not in the document]
Comment (jlusiardi) [reply gatt/gatt_linux.py]: I updated the pull request.

[the remark this answers is not in the document]
Comment (larsblumberg) [reply gatt/gatt_linux.py]: Thank you! Let me get this great feature merged soon.

**Comment by LionNatsu** [comment]
Is it also possible to write a descriptor? I am developing an Android application. There are functions to read and write characteristics and descriptors in Android SDK https://developer.android.com/reference/android/bluetooth/BluetoothGatt .

I want to do this in Python as a light-weight proof-of-concept demo first. It will be great if both reading and writing will be supported in future versions :)

**Comment by Jc2k** [comment]
It would be really handy to have a release with this PR in if anyone can help, thanks!

**Comment by larsblumberg** [comment]
Hi guys, I'm talking to Senic (where I was working and where I wrote this library) as I would like to push an update soon with working Pull Requests.

**Comment by larsblumberg** [comment]
Quick and short update: I met @kelada yesterday who manages Senic's software stack which itself depends on this library. We agreed that I will review and test the open PRs and given they work we can merge them.

We agreed that it's very important to add test coverage to this library before accepting any other new PRs (despite the already opened ones, which will still be merged if functioning). Any support from the community in this direction is welcome, though I will start with the groundwork for test coverage soon.

**Comment by kelada** [comment]
@larsblumberg when you think the PR is ready; ping me and I'll test it against our stack and make sure it's all good on that side and I'll get @jlusiardi  work merged in 👍

**Comment by jlusiardi** [comment]
Hi,
I just wanted to ask for any progress here?
Regards Joachim

**Comment by larsblumberg** [comment]
Sorry for the kind silence. Testing the branches is on my to do list for this weekend. Will keep you updated.

**Comment by jlusiardi** [comment]
Ok cool thanks!

**Comment by larsblumberg** [comment]
@jlusiardi I just tested your Merge Request using Senic's Nuimo Controller. Everything continued to work, nothing breaks (as expected!).

I've tested with BlueZ version 5.44 on a Raspberry Pi model 3 B+.

@kelada I feel very confident that this MR can be safely merged without affecting performance or functionality of the Nuimo Linux SDK.

**Comment by kelada** [comment]
@larsblumberg thank you for this testing, I will try and get this merged in before the end of next week.

**Comment by jlusiardi** [comment]
Thank you!

**Pull request merged**